clients/jack_client.cpp

```diff
@@ -149,30 +153,146 @@ static void done(int sig)
+    if (!sfizz_load_or_import_file(synth.handle(), fpath, &importFormat)) {
+        std::cout << "Could not load the instrument file: " << fpath << '\n';
+        return false;
+    }
+
+    std::cout << "Instrument loaded: " << fpath << '\n';
+    std::cout << "===========================" << '\n';
+    std::cout << "Total:" << '\n';
+    std::cout << "\tMasters: " << synth.getNumMasters() << '\n';
+    std::cout << "\tGroups: " << synth.getNumGroups() << '\n';
+    std::cout << "\tRegions: " << synth.getNumRegions() << '\n';
+    std::cout << "\tCurves: " << synth.getNumCurves() << '\n';
+    std::cout << "\tPreloadedSamples: " << synth.getNumPreloadedSamples() << '\n';
+#if 0 // not currently in public API
+    std::cout << "===========================" << '\n';
+    std::cout << "Included files:" << '\n';
+    for (auto& file : synth.getParser().getIncludedFiles())
+        std::cout << '\t' << file << '\n';
+    std::cout << "===========================" << '\n';
+    std::cout << "Defines:" << '\n';
+    for (auto& define : synth.getParser().getDefines())
+        std::cout << '\t' << define.first << '=' << define.second << '\n';
+#endif
+    std::cout << "===========================" << '\n';
+    std::cout << "Unknown opcodes:";
+    for (auto& opcode : synth.getUnknownOpcodes())
+        std::cout << opcode << ',';
+    std::cout << '\n';
+    if (importFormat) {
+        std::cout << "===========================" << '\n';
+        std::cout << "Import format: " << importFormat << '\n';
+    }
+    // std::cout << std::flush;
+
+    return true;
+}
+
+
+std::vector<std::string> string_tokenize(std::string str) {
+    std::vector<std::string> tokens;
+    std::string part = "";
+    for (size_t i=0; i<str.length(); i++) {
+        char c = str[i];
+        if (c==' ' && part!="") {
+            tokens.push_back(part);
+            part= "";
+        } else if (c== '\"'){
+            i++;
+            while (str[i]!='\"') { part += str[i]; i++; }
+            tokens.push_back(part);
+            part= "";
+        } else {
+            part += c;
+        }
+    }
+    if (part!="") {
+        tokens.push_back(part);
+    }
+    return tokens;
+}
+
+
+void cli_thread_proc() {
+    while (!shouldClose) {
+        std::cout << "\n> ";
+
+        std::string command;
+        std::getline(std::cin, command);
+        std::size_t pos = command.find(" ");
+        std::string kw = command.substr(0, pos);
+        std::string args = command.substr(pos+1);
+        std::vector<std::string> tokens = string_tokenize(args);
+
+        if (kw=="load_instrument") {
+            //args.erase(std::remove(str.begin(), str.end(), '\"'), str.end());
+            load_instrument(tokens[0].c_str());
+        }
+        else if (kw=="set_oversampling") {
+            try {
+                synth.setOversamplingFactor(stoi(args));
+            } catch (...) {
+                std::cout << "ERROR: Can't set oversampling!\n";
```

Comment:
None of these throw, however they're not thread-safe so we need a mutex in the callback. You can check the `SfizzVstProcessor.cpp` file for an example, especially the `doBackgroundWork()` and `process()` functions. Basically you need to fully lock in this `cli_thread_proc()` method, and use `try_lock()` in the real-time thread (and exit if you can't take the lock).